File: 2762.continuous-subarrays.py
#
# @lc app=leetcode id=2762 lang=python3
#
# [2762] Continuous Subarrays
#

# @lc code=start
from collections import deque
from sortedcontainers import SortedList
class Solution:
    def continuousSubarrays(self, nums: list[int]) -> int:
        # max_min = SortedList()
        # left, ans = 0 ,0
        # for i in range(len(nums)):
        #     max_min.add(nums[i])
        #     while max_min[-1] - max_min[0] > 2:
        #         max_min.remove(nums[left])
        #         left += 1

        #     ans += i - left + 1
        # return ans

        #todo This is a faster solution with queue and deque so that we do not need to sort the list to add on the time complexity
        l, res = 0, 0
        minD, maxD = deque(), deque()
        
        for r in range(len(nums)):
            while minD and nums[minD[-1]] >= nums[r]: minD.pop()
            while maxD and nums[maxD[-1]] <= nums[r]: maxD.pop()
            minD.append(r)
            maxD.append(r)
            
            while nums[maxD[0]] - nums[minD[0]] > 2:
                l += 1
                if minD[0] < l: minD.popleft()
                if maxD[0] < l: maxD.popleft()
            
            res += r - l + 1
        
        return res
        
# @lc code=end

File: 2762.continuous-subarrays.java
/*
 * @lc app=leetcode id=2762 lang=java
 *
 * [2762] Continuous Subarrays
 */
import java.util.*;
// @lc code=start
class Solution {
    public long continuousSubarrays(int[] nums) {
        //使用两个单调队列来维护滑窗内部的最大值和最小值 同时方便滑窗更新

        //最重要的是题目条件等价于维护窗口内最大值和最小值的差值小于等于2
        Deque<Integer> minQ = new ArrayDeque<>();
        Deque<Integer> maxQ = new ArrayDeque<>();
        long ans = 0;
        int left = 0;
        for (int right = 0; right < nums.length; right++) {
            while (!minQ.isEmpty() && minQ.peekLast() > nums[right]) {
                minQ.pollLast();
            }
            minQ.offerLast(nums[right]);
            while (!maxQ.isEmpty() && maxQ.peekLast() < nums[right]) {
                maxQ.pollLast();
            }
            maxQ.offerLast(nums[right]);
            while (maxQ.peekFirst() - minQ.peekFirst() > 2) {
                if (minQ.peekFirst() == nums[left]) {
                    minQ.pollFirst();
                }
                if (maxQ.peekFirst() == nums[left]) {
                    maxQ.pollFirst();
                }
                left++;
            }
            ans += right - left + 1;
        }
        return ans;

    }
}
// @lc code=end

